source/slang/slang-ir-validate.cpp

```diff
@@ -223,6 +223,15 @@ namespace Slang
         {
         case kIROp_DifferentiableTypeDictionaryItem:
             return;
+        default:
+
+            // It is possible that a global parameter will not emit their ancestor until after
```

Comment:
I don't quite understand why we are hitting a new exception scenario that is not covered previously, and I am not sure this is the correct workaround. Can you describe in detail what are the issue we are working around here?

Reply:
The workaround to emit `in` parameters directly where called as arguments relies on the following steps:
```
                // Add an entry point parameter for all the inputs.
                auto firstBlock = entryPointFunc->getFirstBlock();
                builder.setInsertInto(firstBlock);
                auto inputParam = builder.emitParam(inputStructType);
                builder.addLayoutDecoration(inputParam, paramLayout);
```
emits a kIROp_Param for each eventual `GlobalParam`. 

since these kIROp_Param are the actual GlobalParam and are treated as such in the code, we can do the same here
```
                ...
                // Replace all global variable references with
                // actual global parameter for all Call's
                for (Index i = 0; i < inputVars.getCount(); i++)
                {
                            ...
                            auto field = builder.emitFieldExtract(inputType, inputParam, inputKeys[i]);
                            ...
```

The downside of doing this is that during `validateIRInstOperand` the kIROp_Param will not be in scope since it is a entryPoint parameter (different function). Only until later in compilation are all entryPoint's transformed into a GlobalParam in global scope.

Reply:
I think we should avoid changing the general IR validation logic just to allow us violating it, even if we eventually are correcting it, because this is not a scalable design. If somebody adds in a pass between where we violate the rule and where we correct the IR, that pass can fail/crash for reasons suprising to the author. Is there another way to get things to work without this exception?

Reply:
It seems that this replacement step kicks in too early. We can consider moving it to after the parameter is turned into a global parameter again.